wikimedia/wikipedia-ios
# Determine new notifications in service extension and display

**Phabricator:** https://phabricator.wikimedia.org/T287310
(also a little bit of https://phabricator.wikimedia.org/T287033 & https://phabricator.wikimedia.org/T285353)

### Notes
This extends the work done in #4026 to pull the latest unread push notifications from the notifications API, filters out previously seen notifications via an app container cache file and displays the results. If the result is a single new unread notification, it modifies the push content to display it's `header` value. Otherwise it defaults to the "New activity on Wikipedia" string.

After review, please wait for #4026 to be reviewed and merged. Then update this with feature/notifications and merge into that feature branch.

These steps can be made on production in Test Wikipedia or on the Beta Cluster via the Staging scheme. Just make sure you're triggering notifications at the correct site on desktop (see step 4)

### Test Steps
1. Fresh install from this branch and log in from an account that has all checkboxes checked on desktop Wikipedia notifications preferences. Go to Settings > Notifications and enable notifications. After allowing permissions, watch the traffic via a proxy for the subscription call. Use the `providertoken` and `topic` values from this call in the push notification utility.
2. Send yourself a push notification via the push notification utility.
3. This first one's content will likely default to "New activity on Wikipedia", unless you happen to have a single notification triggered to yourself in the last 10 minutes.
4. From another account on desktop Wikipedia, post a message to your device account's user talk page at https://test.wikipedia.org/wiki/User_talk:{name} or https://en.wikipedia.beta.wmflabs.org/wiki/User_talk:{name}
5. Trigger a push notification to your device again via the push utility. Confirm you see a new notification about someone posting to your talk page.
6. Do something again on desktop Wikipedia to trigger a notification to your device account (could be thanking an edit, reverting an edit, posting to their talk page again, etc.).
7. Confirm you see a new notification with proper content based on what notification you triggered in step 6.

Note for step 5 and 7, we are displaying the "header" value from the Notifications API as the content, to match the default notification requirement in https://phabricator.wikimedia.org/T287033.
